// space-lift/commonjs/lift.js
"use strict";
exports.__esModule = true;
var iterator_1 = require("./iterator");
var lift = function (obj) {
    if (obj instanceof Array)
        return new exports.ArrayOps(obj);
    if (obj instanceof Date)
        return new exports.DateOps(obj);
    if (typeof obj === 'string')
        return new exports.StringOps(obj);
    if (typeof obj === 'number')
        return new exports.NumberOps(obj);
    return new exports.ObjectOps(obj);
};
exports["default"] = lift;
function getValue(input) {
    return input && input['_isLiftWrapper']
        ? input.value()
        : input;
}
exports.getValue = getValue;
function makeOps() {
    var Ops = /** @class */ (function () {
        function Ops(_value) {
            this._value = _value;
            this._isLiftWrapper = true;
        }
        Ops.prototype.value = function () { return this._value; };
        return Ops;
    }());
    Ops.prototype[iterator_1.iteratorSymbol] = iterator_1.singleValueIterator(function (self) { return self._value; });
    return Ops;
}
exports.ArrayOps = makeOps();
exports.ArrayOps.prototype[iterator_1.iteratorSymbol] = function () {
    var i = 0;
    var self = this;
    return {
        next: function () {
            return i < self._value.length
                ? { value: self._value[i++], done: false }
                : { done: true };
        }
    };
};
exports.ObjectOps = makeOps();
exports.NumberOps = makeOps();
exports.StringOps = makeOps();
exports.DateOps = makeOps();
